api: Prevent special characters in streams name.

madrix01 opened the pull request:
Special characters, including `\r`, `\n`, and more esoteric codepoints
like non-characters, can negatively affect rendering and UI behavior,
such as creating two streams with the same name, but with different URLs.

Check for, and prevent making new streams with, characters in the
Unicode categories of `Cc` (control characters), `Cs` (surrogates), and
`Cn` (unassigned, non-characters)

To learn more about Unicode characters 
- [Unicode Wikipedia](https://en.wikipedia.org/wiki/Unicode)
- https://www.unicode.org/reports/tr44/#General_Category_Values
- [Examples of Surrogate category](https://www.fileformat.info/info/unicode/category/Cs/list.htm) 
- [Examples of Non characters](https://www.unicode.org/faq/private_use.html#nonchar4)

For tests, I have created test for `Cc` and `Cn` characters, as python don't allow surrogate characters.

Fixes a part of #20128

madrix01 commented:
@alexmv can you please review this and suggest any changes required, Thank you!!

alexmv commented on zerver/lib/streams.py:
```diff
@@ -185,6 +187,10 @@ def check_stream_name(stream_name: str) -> None:
                 _("Stream name '{}' contains NULL (0x00) characters.").format(stream_name)
```
This is covered by `Cc`, below.  It's a more specific error message, but I don't know that it merits still having its own check.

timabbott replied:
I think we can remove this line and just keep the existing test (with the new more generic string).  

We added this because it appeared in the wild, but there's no reason to have a duplicated check for it now that we're doing something better.

madrix01 replied:
Should we mention the category or character in the error message

alexmv replied:
Putting the character in the error message is nice, in that it's specific about what needs to be changed -- but bear in mind that we'd need to come up with some way to escape them.  Otherwise we'd say:

```
"Invalid character 
 in stream name!"
```

...which might be confusing.  And while [it's not prohibited to send back a non-character](https://www.unicode.org/faq/private_use.html#nonchar7) if that's what we got sent, it'd be nice to send it as `\xFFFF` and not `￿`.

So, if we can do it carefully, it's a fine idea.  Note that it will only show the _first_ offending character, not all of them.

madrix01 replied:
@alexmv @timabbott  I think we stringify `Cc` characters but as you mentioned non-characters aren't sent back, and besides we may also face problems with surrogate characters too if that case ever happens. 
I think `Invalid unicode characters in stream name!` would be a better error message. What are your views about this?

alexmv replied:
> I think we stringify `Cc` characters

I'm not sure what you mean by "stringify" here.  We don't currently send back the offending character in any form, so I don't think there's anything to say about what we currently do with `Cc` characters.

> as you mentioned non-characters aren't sent back

We do _currently_ (before this PR) faithfully use any instances of `\xFFFF`, like any other character.  I was talking about if it was going to be valid to send the characters back in an error message.  The link I pointed to said doing so was acceptible.

> and besides we may also face problems with surrogate characters too if that case ever happens

I strongly suspect that we won't actually be able to get half of a surrogate pair this far down the stack, since that requires UTF-16 encoding.  One point is that UTF-8 can't actually encode codepoints that are surrogate pairs, so I don't know how we'd get them here.  But we certainly wouldn't be able to send them back in the error message, yes.

> I think `Invalid unicode characters in stream name!` would be a better error message.

I don't think we should say they're "unicode characters" -- they're just characters.  And I think I made my opinion clear in my previous message -- if we can carefully tell the user the offending character, that would be useful.

I guess one simpler alternative would be to tell them which character it is by _position_, instead of sending it back.  So have the error be `Invalid character in stream name, at position {x}!`

alexmv commented on zerver/tests/test_subs.py:
```diff
@@ -185,6 +184,37 @@ def test_creating_streams(self) -> None:
 
+    def test_invalid_stream_name(self) -> None:
+        """
+        Creating a stream with invalid 'Cc', 'Cs' and 'Cn' category of unicode characters in stream name
```
The test only tests `Cc` and `Cn`.  Which is fine, just we shouldn't claim more than we're actually doing. :)

madrix01 replied:
Sorry for that, I removed it thank you !!

alexmv reviewed (commented):
This is great, @madrix01!  I have a couple small nitpicks, but otherwise this is looking pretty close.

cc @mateuszmandera who was looking at the migration to follow this up with to fix any past instances.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -185,6 +187,10 @@ def check_stream_name(stream_name: str) -> None:
             )
 
+        unicodeCategory = unicodedata.category(i)
+        if unicodeCategory in ["Cc", "Cs", "Cn"]:
```
We use underscores, not camelCase, for variable names.

madrix01 commented on zerver/tests/test_subs.py:
```diff
@@ -123,7 +123,6 @@ def test_creating_streams(self) -> None:
         events: List[Mapping[str, Any]] = []
         with self.tornado_redirected_to_list(events, expected_num_events=1):
             ensure_stream(realm, "Public stream", invite_only=False, acting_user=None)
-
```
Done!

mateuszmandera commented on zerver/lib/streams.py:
```diff
@@ -175,15 +176,15 @@ def create_streams_if_needed(
-                _("Stream name '{}' contains NULL (0x00) characters.").format(stream_name)
-            )
+        unicode_category = unicodedata.category(i)
+        if unicode_category in ["Cc", "Cs", "Cn"]:
```
Hmm, shouldn't we extract a helper function for validating these characters, to avoid duplicating the same logic with b3c58f454f0dce8a88c696305945bac41f9786bc?

madrix01 replied:
@timabbott @mateuszmandera This seems like a good idea, should we add this in `zerver/lib/validator.py`?

timabbott replied:
I'd put it in a new `zerver/lib/string_validation.py`; `validator.py` is a library with a very specific purpose.

alexmv replied:
When you're doing this refactoring, zerver/migrations/0371_invalid_characters_in_topics.py has an additional use of this pattern now as well.

timabbott replied:
Yes, though we should just add a `# Copy of {function_name}` comment there rather than having it call into the new function; since if we change the set of character classes, we need to repeat the migration, rather than changing one that has already been run on some systems.

madrix01 replied:
@timabbott @alexmv If we create a common function for string validation, how can we generalize the error messages, Like we should pass it with the arguments of function created in `zerver/lib/string_validation.py`
```python
def check_invalid_string(var: str, error: str) -> None:
```
or else we just pass the object name such as `stream name`, `topic name` or `user group name`. 
```python
raise JsonableError(_("Invalid unicode characters in {}").format(object_name))
```

mateuszmandera replied:
@madrix01 Another alternative could be to raise e.g. `ValidationError` in the character validation function, and the caller can catch it and re-raise its own JsonableError.

madrix01 replied:
@mateuszmandera then we would have to add conditionals for different types of errors, in different files, is it right? Can you give an example?

madrix01 commented:
@timabbott @alexmv I have created a new library `check_input_string` in `zerver/lib/string_validation.py`. And replaced this for `stream name`, and also added some tests. Changed some error messages as suggested by @alexmv.

alexmv commented:
We just discovered that under the `unicodedata` distributed with Python 3.6, some Emoji are classified as `Cn`, and no `So`:
```
$ unicode 1f929 --long
U+1F929 GRINNING FACE WITH STAR EYES
UTF-8: f0 9f a4 a9 UTF-16BE: d83edd29 Decimal: &#129321; Octal: \0374451
🤩
Category: So (Symbol, Other); East Asian width: W (wide)
Unicode block: 1F900..1F9FF; Supplemental Symbols and Pictographs
Bidi: ON (Other Neutrals)

$ python3.6 -c 'import unicodedata; print(unicodedata.category("\U0001f929"))'
Cn

$ python3.7 -c 'import unicodedata; print(unicodedata.category("\U0001f929"))'
So
```

Given that ,we should drop `Cn` from the list that we exclude.   I think ideally we'd refuse anything containing one of the 66 explicit non-characters, which are never going to change, but accept anything which is unassigned -- that way we're forward-compatible with future versions of Unicode.

alexmv commented:
Also, as a follow-up PR, we should strip any invalid characters from incoming emails via the incoming email gateway.  Otherwise, emails with invalid characters in their `Subject` will get dropped, and generate a `JsonableError` in the incoming mail gateway codepath; see https://sentry.io/share/issue/104c521c822d4c27ba41fa50185a7f99/ for such a stacktrace.

madrix01 commented:
@alexmv In the latest commit, I have set a check for all the 66 non-characters. Is this the correct implementation? Thank you!!

madrix01 commented:
@timabbott @alexmv I have made the changes for the 66 non-characters, is this the correct way? I was confused about sending the error message from the common function, for which I have made a fix, where we specify the object for which we are validating like `Stream name`, `Topic name` etc. Can this work?

alexmv commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,93 @@
+    "\uFDEB",
+    "\uFDEC",
+    "\uFDED",
+    "\uFDEE",
+    "\uFDEF",
+    "\uFFFE",
+    "\uFFFF",
+    "\U0001FFFE",
+    "\U0001FFFF",
+    "\U0002FFFE",
+    "\U0002FFFF",
+    "\U0003FFFE",
+    "\U0003FFFF",
+    "\U0004FFFE",
+    "\U0004FFFF",
+    "\U0005FFFE",
+    "\U0005FFFF",
+    "\U0006FFFE",
+    "\U0006FFFF",
+    "\U0007FFFE",
+    "\U0007FFFF",
+    "\U0008FFFE",
+    "\U0008FFFF",
+    "\U0009FFFE",
+    "\U0009FFFF",
+    "\U0010FFFE",
+    "\U0010FFFF",
+    "\U000AFFFE",
+    "\U000AFFFF",
+    "\U000BFFFE",
+    "\U000BFFFF",
+    "\U000CFFFE",
+    "\U000CFFFF",
+    "\U000DFFFE",
+    "\U000DFFFF",
+    "\U000EFFFE",
+    "\U000EFFFF",
+    "\U000FFFFE",
+    "\U000FFFFF",
+]
```
This is very verbose -- since there's a simple pattern behind it, we can probably make it more understandable and concise.  And magic constants like this are the things that benefit greatly from a comment pointing to their source -- otherwise, they just leave me wondering "why is it _this_ set of of things?"

```suggestion
# There are 66 Unicode non-characters; see 
# https://www.unicode.org/faq/private_use.html#nonchar4
unicode_non_chars = [
    chr(x)
    for x in list(range(0xFDD0, 0xFDF0)) # FDD0 through FDEF, inclusive
    + list(range(0xFFFE, 0x110000, 0x10000))  # 0xFFFE, 0x1FFFE, ... 0x10FFFE inclusive
    + list(range(0xFFFF, 0x110000, 0x10000))  # 0xFFFF, 0x1FFFF, ... 0x10FFFF inclusive
]

alexmv commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,93 @@
+]
+
+
+def check_invalid_string(var: str, entity: str, max_length: int) -> None:
+    if var.strip() == "":
+        raise JsonableError(_("{} can't be empty!").format(entity))
```
This function name feels really not-right.  "valid" is one of those adjectives which is really squirrelly -- it doesn't tell me much information unless it's universally clear what a valid or invalid thing of its class is.  For example, "valid number" is pretty ambiguous, since that could potentially be integers, floats, complex numbers, etc.  "valid string" is similarly pretty unclear -- valid for _what_?  When should I reach for this function?

There's also another smell here that we're barking up the wrong tree doing it this way -- there's no way to translate this error.  For example, French translators will translate "{} can't be empty!" as "{} ne peut pas être vide!" which, if thrown, will produce the polyglot "Stream name ne peut pas être vide!" and not "Le nom de canal ne peut pas être vide!"

I think this is an example of [the wrong abstraction](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction); both the inability to have a good name for it, and the inability to return sensible error messages, point that we're trying too hard to reduce duplication, and it's making it _harder_ to write clear code.

I think the right thing here is to have a function for stream names, and a function for instance names.  Put them in the same file so they can both use `unicode_non_chars`, and so their parallelism is evident, but making them their own functions means their errors can be very straightforward, they won't have extraneous flags, and it'll be clear when to call them.

If we really want to de-duplicate something, I think the only part of this that we could pull out is a predicate (function that returns a `bool`) that checks if a character is in `Cc`, `Cs`, or a non-character.

madrix01 commented:
@alexmv As per your review, I have removed the hardcoded list of non-characters, besides, I made a common function checking of Unicode characters of `Cc`, `Cs`, and `non-characters` which returns the position of the character if present or `-1`, and introduced `check_stream_name` to avoid the `wrong abstraction`, and will add `check_topic_name` function in a future commit.

mateuszmandera commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,37 @@
+import unicodedata
```
Do we actually want a new file for these, rather than putting them in `zerver/lib/validators.py`?

alexmv replied:
I believe that @timabbott was arguing for its own file earlier in this PR: https://github.com/zulip/zulip/pull/20375#discussion_r766040701

mateuszmandera commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,37 @@
+]
+
+
+def check_invalid_unicode_character(var: str) -> int:
```
It looks to me like the function should  return `bool` rather than `int`, since it's communicating one of two possible values.

To make the function name clearer, should it perhaps be `check_string_for_disallowed_characters` or something like that? Perhaps Alex will have a comment on that, given the discussion about the term "invalid X" earlier.

mateuszmandera replied:
Also, we generally want to avoid abbreviations in variable names (though functions in `validator.py` seem to mostly not be following that :neutral_face: ) - should we just call the argument `string` perhaps?

mateuszmandera commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,37 @@
+]
+
+
+def check_invalid_unicode_character(var: str) -> int:
```
Do we need this to return the position of the first invalid character rather than just `True/False`?

To make the function name clearer, should it perhaps be `check_string_for_disallowed_characters` or something like that? Perhaps Alex will have a comment on that, given the discussion about the term "invalid X" earlier.

alexmv replied:
"disallowed" it a little in the direction of "invalid" in that it doesn't tell me much about why.  Maybe `check_string_is_printable` ?  But we're wandering a little into yak-shaving territory.

madrix01 replied:
> Do we need this to return the position of the first invalid character rather than just `True/False`?

Yes, because we are also specifying the index at which the character is found.

aw5631 commented:
@zulipbot comment

On Fri, 31 Dec 2021, 19:42 Mateusz Mandera, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In zerver/lib/string_validation.py
> <https://github.com/zulip/zulip/pull/20375#discussion_r777049325>:
>
> > +import unicodedata
> +
> +from django.utils.translation import gettext as _
> +
> +from zerver.lib.exceptions import JsonableError
> +from zerver.models import Stream
> +
> +unicode_non_chars = [
> +    chr(x)
> +    for x in list(range(0xFDD0, 0xFDF0))  # FDD0 through FDEF, inclusive
> +    + list(range(0xFFFE, 0x110000, 0x10000))  # 0xFFFE, 0x1FFFE, ... 0x10FFFE inclusive
> +    + list(range(0xFFFF, 0x110000, 0x10000))  # 0xFFFF, 0x1FFFF, ... 0x10FFFF inclusive
> +]
> +
> +
> +def check_invalid_unicode_character(var: str) -> int:
>
> Do we need this to return the position of the first invalid character
> rather than just True/False?
>
> To make the function name clearer, should it perhaps be
> check_string_for_disallowed_characters or something like that? Perhaps
> Alex will have a comment on that, given the discussion about the term
> "invalid X" earlier.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/20375#pullrequestreview-842295928>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADRP6AI2SJ4A2R3PXSBDBQTUTYBQRANCNFSM5I34MXYA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

alexmv commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,37 @@
+        unicode_character = unicodedata.category(var[i])
+        if (unicode_character in ["Cc", "Cs"]) or var[i] in unicode_non_chars:
+            return i
+    return -1
```
I'm not a big fan of -1 as an error value, because it's easy to forget to check for.  Making this be typed `Optional[int]` and return `None` would mean mypy would helpfully force us to enforce that we checked for `is not None` before we can do anything with it that assumes it's an int.

Looking at how it's currently called, I guess that wouldn't help here, but I'd prefer to avoid the -1 convention.

madrix01 replied:
I tried the same earlier but was facing a problem in using `Optional`, it is now resolved.

madrix01 commented:
@alexmv I have implemented the new changes, can you please have a look?

alexmv reviewed (commented):
A couple comments inline, and the commit message need to be updated.

Thinking about what to put in the commit message made me realize that the single commit here isn't quite how I would do it personally.  I don't think this is a requirement, but I would probably split this into the following commits:
 - Move `check_stream_name` into `string_validation.py`, unchanged.
 - Move `validate_topic` into `string_validation.py`, rename to `check_stream_topic`, otherwise unchanged.
 - Add `check_string_is_printable` and `unicode_non_chars`, and switch `check_stream_topic` to use them.  The commit message clarifies that this fixes things to re-allow topics to have some emoji that are characterized as `Cn` in Python 3.6.
 - Switch `check_stream_name` to use `check_string_is_printable`.
 - Copy `check_string_is_printable` into `zerver/migrations/0371_invalid_characters_in_topics.py` and change it to use that definition.

Part of why is that it feels a little odd about this commit as-is is that we're introducing this abstraction without doing anything about the one we introduced in b3c58f454f0dce8a88c696305945bac41f9786bc.  I know that you're planning to come back to that, but grouping the commits as above unifies similar things next to each other, then fixes a bug in the old implementation, then improves the other, then cleans up the last location which we can't unify -- which seems like a more coherent story arc to me.

I know you've been working on this PR for a while, so the above is a suggestion, and not a requirement.  But I hope it's a compelling one!

madrix01 commented on zerver/tests/test_subs.py:
```diff
@@ -3007,9 +3007,7 @@ def test_stream_name_contains_null(self) -> None:
-        self.assert_json_error(
-            result, f"Stream name '{stream_name}' contains NULL (0x00) characters."
-        )
+        self.assert_json_error(result, "Invalid character in stream name, at position 3!")
```
Yes this seems readable and can be easily understood, I have changed it in the new commit.

madrix01 commented on zerver/lib/string_validation.py:
```diff
@@ -0,0 +1,39 @@
+from zerver.lib.exceptions import JsonableError
+from zerver.models import Stream
+
+unicode_non_chars = [
+    chr(x)
+    for x in list(range(0xFDD0, 0xFDF0))  # FDD0 through FDEF, inclusive
+    + list(range(0xFFFE, 0x110000, 0x10000))  # 0xFFFE, 0x1FFFE, ... 0x10FFFE inclusive
+    + list(range(0xFFFF, 0x110000, 0x10000))  # 0xFFFF, 0x1FFFF, ... 0x10FFFF inclusive
+]
```
Sorry, I missed it, I have added it. Thank you !!

madrix01 commented:
@alexmv thanks for the review. This issue has taken a lot of turns and has solved a wider range of issues and removed code duplication with string validation and one commit can't really describe the changes. I too think making different commits for different changes would be a better idea. For now I have pushed a single commit with updated commit message, added `check_topic_name`, and updated the tests.
For making different commits should I make a new PR so that it would be convenient?

alexmv commented:
Thanks for all of this work, @madrix01!  I split apart the commits, and merged most of it (all but the last commit) into `main` as df50280c5423f44a9cf6bfdabfed85a5cef32dc1.  We're a little nervous to get df50280c5423f44a9cf6bfdabfed85a5cef32dc1 merged sooner rather than later, because that will unbreak people, in terms of being able to use emoji in topic names.

I didn't merge the last commit because we also need to write a migration for this.  If we merge the last commit, we could have folks with existing "invalid" stream names, who might get errors when trying to send to them, or edit properties of them.  We should merge this last commit that prevents making _new_ ones at the same time as we clean up the _old_ ones.

The migration will be similar to `zerver/migrations/0371_invalid_characters_in_topics.py`, but:
 - We don't need to do the fancy batching -- we can just get the `Stream` model and iterate over `Stream.objects.all()` and update them via `stream.save()`
 - Any stream name which ends up empty needs to get replaced with something probably-unique -- like `Unknown stream {stream.id}` or similar.
 - Any stream which isn't unique after removing invalid characters needs to be made unique, possibly via appending ` (#{stream.id})` or similar.

madrix01 commented:
@alexmv thanks for guiding me while solving this issue. Can I work on creating migration for the stream name as you mentioned?

alexmv commented:
@madrix01: Let's keep the changes in one PR, since we want to merge them at the same time?  I'll leave some comments there, but please close than one, and re-push this with that commit added to it.

mateuszmandera commented:
The migration duplicates work with https://github.com/zulip/zulip/pull/20522, aside of #20522 still having an older version of the logic for `character_is_printable`. Its main upsides are going realm-by-realm instead of fetching the entirety of `Stream.objects.all()` (which I'm not sure if is large enough on Zulip Cloud to be an issue?) and some somewhat useful printing to see the progress.

Should we perhaps take the migration logic from it and plug this PR's  `character_is_printable` into it?

alexmv commented:
@mateuszmandera: Ah, I'd forgotten (or not seen) the migration you made!  And I think I prefer that one, albeit with the `character_is_printable` from here, yes.

My ideal is that we land these two changes back-to-back.  If you adjust your PR, we can merge it immediately following the first commit of this one.

@madrix01, I'm sorry for the duplicated work!  I intend to close #20782, since regardless of if we use your migration, that work is contained in this PR now.

mateuszmandera commented:
@alexmv Okay, I've updated the PR with the `character_is_printable` logic

aw5631 commented:
Okay good, now suck my shlong.

On Thu, 20 Jan 2022, 15:46 Mateusz Mandera, ***@***.***>
wrote:

> @alexmv <https://github.com/alexmv> Okay, I've updated the PR with the
> character_is_printable logic
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/20375#issuecomment-1017642066>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADRP6APVNUHVPJBGOQAO4Z3UXAU3VANCNFSM5I34MXYA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

alexmv commented:
I think this, in conjunction with #20522, is in good shape!

Can you rebase this?  With the drop of 18.04 support, the testsuite names have changed, so Github wants to have the new tests run before it will let it merge.

@mateuszmandera: #20522 will also need a rebase, it looks like.  I'll merge them back-to-back once they're rebased.

madrix01 commented:
@alexmv the commit that is yet to be merged is df50280. right?

madrix01 commented:
@alexmv the commit that is yet to be merged is df50280. right?

alexmv commented:
@madrix01: No, df50280c5423f44a9cf6bfdabfed85a5cef32dc1 is the one we _have_ merged already.  If you click that link, under the commit message, you'll even see a little `main` label which tells you which branches it's in.

When you rebase, you'll want to keep the first commit, and not the second -- we'll pull #20522 for the migration.

madrix01 commented:
@alexmv I'm facing problem with git as I've formatted my laptop. 
![image](https://user-images.githubusercontent.com/51414879/151673133-439b2366-60cb-48b1-9d5a-81149154776d.png)
it is showing `#empty` beside the commit.

madrix01 commented:
And on `git status` :
```
On branch valid_stream_name
Your branch is ahead of 'origin/valid_stream_name' by 231 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
```

mateuszmandera commented:
> When you rebase, you'll want to keep the first commit, and not the second -- we'll pull #20522 for the migration.

@alexmv Hmm, it looks good to me without rebasing, not sure if I'm missing something? The migration number `0375` looks correct with the current `upstream/main`

alexmv commented:
@mateuszmandera: Left a comment on that PR.

@madrix01: It looks like the re-push inhttps://github.com/zulip/zulip/pull/20375#event-5899226309 from cd1117d to bd7a1d1 made the second commit into a no-op empty commit.  Something must have gone amiss with that rebase.  I've re-applied the commit in cd1117d454959ea71df6caa8e70a950ede222cd4 and re-pushed that.

mateuszmandera commented:
Ah, thanks for fixing, were just working with @madrix01 to recover that as well! 😁

alexmv commented:
Ah, it looks like the commit I re-pushed included the migration.  Repushed with that trimmed out.

alexmv commented:
Merged as d1234ef18c.
